cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/InMemoryOperationRuntime.java

```diff
@@ -0,0 +1,199 @@
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nullable;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link OperationRuntime} implementation when operations are running inside same service.
+ */
+public class InMemoryOperationRuntime extends AbstractIdleService implements OperationRuntime {
+
+  private final CConfiguration cConf;
+  private final ReadWriteLock runtimeLock;
+  private final Map<OperationRunId, OperationController> controllers;
+  private final OperationRunner runner;
+  private final OperationStatePublisher statePublisher;
+  private final TransactionRunner transactionRunner;
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryOperationRuntime.class);
+
+  @Inject
+  InMemoryOperationRuntime(CConfiguration cConf, OperationRunner runner,
+      OperationStatePublisher statePublisher,
+      TransactionRunner transactionRunner) {
+    this.cConf = cConf;
+    this.runner = runner;
+    this.statePublisher = statePublisher;
+    this.transactionRunner = transactionRunner;
+    this.runtimeLock = new ReentrantReadWriteLock();
+    this.controllers = new HashMap<>();
+  }
+
+  /**
+   * Runs an operation given the detail.
+   *
+   * @param runDetail detail of the run
+   * @return {@link OperationController} for the run
+   */
+  public OperationController run(OperationRunDetail runDetail) {
+    OperationController controller = getController(runDetail);
+    if (controller != null) {
+      LOG.debug("Operation has already been started: {}", runDetail.getRunId());
+      return controller;
+    }
+    try {
+      updateController(runDetail.getRunId(), runner.run(runDetail));
+    } catch (IllegalStateException e) {
+      statePublisher.publishFailed(runDetail.getRunId(),
+          new OperationError(e.getMessage(), Collections.emptyList())
+      );
+    }
+    return controllers.get(runDetail.getRunId());
+  }
+
+  /**
+   * Get controller for a given {@link OperationRunId}.
+   */
+  @Nullable
+  public OperationController getController(OperationRunDetail detail) {
+    Lock lock = this.runtimeLock.readLock();
+    lock.lock();
+    try {
+      return controllers.get(detail.getRunId());
+      // TODO(samik) fetch from store for remote operations
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    RetryStrategy retryStrategy =
+        RetryStrategies.fromConfiguration(cConf, Constants.Service.RUNTIME_MONITOR_RETRY_PREFIX);
+
+    TransactionRunners.run(
+        transactionRunner, context -> {
+          Retries.runWithRetries(
+              () -> {
+                processStartingOperations(context);
```

Comment:
I don't think this service is the right place to do this, as nothing else in this class is handling state transitions. It also means if there is another implementation that would also need to do this? It seemed fine before in the notification subscriber, just need it in the top level service and not the per-topic partition service.

Reply:
The plan was to move it to a abstract base class when new implementation is created. The only issue I think with moving it to subscriber is ho to handle existing running/ stopping operations differs between different runtime implementations